module/mempool/stdmap/chunk_requests.go

```diff
@@ -52,13 +52,24 @@ func (cs *ChunkRequests) ByID(chunkID flow.Identifier) (*verification.ChunkDataP
 // The last boolean parameter returns whether a chunk request for this chunk ID
 // exists in memory-pool.
 func (cs *ChunkRequests) RequestHistory(chunkID flow.Identifier) (uint64, time.Time, time.Duration, bool) {
-	entity, exists := cs.Backend.ByID(chunkID)
-	if !exists {
-		return 0, time.Time{}, time.Duration(0), false
-	}
-	request := toChunkRequestStatus(entity)
+	var lastAttempt time.Time
+	var retryAfter time.Duration
+	var attempts uint64
 
-	return request.Attempt, request.LastAttempt, request.RetryAfter, true
+	err := cs.Backend.Run(func(backdata map[flow.Identifier]flow.Entity) error {
```

Comment:
to fix the race condition between this method and `UpdateRequestHistory`.

Reply:
How does this fix the race condition? Both this and `Backend.ByID` acquire the lock.

Reply:
I think the race condition is between `IncrementAttempt` and `RequestHistory` .

In the previous approach, when `RequestHistory` got the `request` object from `ByID`, it inspects the `Attempt` field, meanwhile `IncrementAttempt` could concurrently updating the `Attempt` field, causing race condition.

If we use `Backend.Run` instead, then `IncrementAttempt` will not be blocked waiting until `RequestHistory` is done with inspecting the fields.

Reply:
`ByID` returns the reference (entity) to the chunk request status, and _releases the lock_ right upon return. So the following piece of code is not protected by any lock. Let's call this part _"unprotected"_:

```
	if !exists {
		return 0, time.Time{}, time.Duration(0), false
	}
	request := toChunkRequestStatus(entity)

	return request.Attempt, request.LastAttempt, request.RetryAfter, true
```

The race condition happens when the request for the same chunk ID is updated _concurrently_ by the `UpdateRequestHistory` while the `RequestHistory` is in its _unprotected_ part (though its `ByID` call returned). Since the `ByID` released the lock, the `UpdateRequestHistory` can obtain the lock immediately and starts updating the request for the same chunk ID that is _read_ by the _unprotected_ part of concurrent `RequestHistory` invocation -> race condition. I examined this hypothesis by `-race` flag as well. 

Now by using the `Run` method, we encapsulate the _entire body_ of `RequestHistory` logic by a mutex lock. So, no `UpdateHistory` can get through while a `RequestHistory` call is in process.

Reply:
> I think the race condition is between IncrementAttempt and RequestHistory .

In current code `IncrementAttempt` is never used.

Reply:
Another idea I thought this morning is to add a RWLock to `Request` object, so that we can use `ByID` to query the Request object, and use a Read lock to read the fields. 

The good part is that we no longer need to lock the entire Backend if we only need to inspect one Request.